// ngx-webstorage/esm5/lib/module.js
import { __decorate, __metadata, __param, __read, __spread } from "tslib";
import { APP_INITIALIZER, Inject, InjectionToken, NgModule, Optional } from '@angular/core';
import { LocalStorageProvider, SessionStorageProvider } from './core/nativeStorage';
import { Services } from './services/index';
import { Strategies } from './strategies/index';
import { StrategyIndex } from './services/strategyIndex';
import { StorageKeyManager } from './helpers/storageKeyManager';
import * as ɵngcc0 from '@angular/core';
export var LIB_CONFIG = new InjectionToken('ngx_webstorage_config');
export function appInit(index) {
    index.indexStrategies();
    return function () { return StrategyIndex.index; };
}
var NgxWebstorageModule = /** @class */ (function () {
    function NgxWebstorageModule(index, config) {
        if (config)
            StorageKeyManager.consumeConfiguration(config);
        else
            console.error('NgxWebstorage : Possible misconfiguration (The forRoot method usage is mandatory since the 3.0.0)');
    }
    NgxWebstorageModule_1 = NgxWebstorageModule;
    NgxWebstorageModule.forRoot = function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: NgxWebstorageModule_1,
            providers: __spread([
                {
                    provide: LIB_CONFIG,
                    useValue: config,
                },
                LocalStorageProvider,
                SessionStorageProvider
            ], Services, Strategies, [
                {
                    provide: APP_INITIALIZER,
                    useFactory: appInit,
                    deps: [StrategyIndex],
                    multi: true
                }
            ])
        };
    };
    var NgxWebstorageModule_1;
    NgxWebstorageModule.ctorParameters = function () { return [
        { type: StrategyIndex },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [LIB_CONFIG,] }] }
    ]; };
    NgxWebstorageModule = NgxWebstorageModule_1 = __decorate([ __param(1, Optional()), __param(1, Inject(LIB_CONFIG)),
        __metadata("design:paramtypes", [StrategyIndex, Object])
    ], NgxWebstorageModule);
NgxWebstorageModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxWebstorageModule });
NgxWebstorageModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxWebstorageModule_Factory(t) { return new (t || NgxWebstorageModule)(ɵngcc0.ɵɵinject(StrategyIndex), ɵngcc0.ɵɵinject(LIB_CONFIG, 8)); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxWebstorageModule, [{
        type: NgModule,
        args: [{}]
    }], function () { return [{ type: StrategyIndex }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [LIB_CONFIG]
            }] }]; }, null); })();
    return NgxWebstorageModule;
}());
export { NgxWebstorageModule };
