// ngx-webstorage/esm5/lib/core/strategyCache.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as ɵngcc0 from '@angular/core';
var StrategyCacheService = /** @class */ (function () {
    function StrategyCacheService() {
        this.caches = {};
    }
    StrategyCacheService.prototype.get = function (strategyName, key) {
        return this.getCacheStore(strategyName)[key];
    };
    StrategyCacheService.prototype.set = function (strategyName, key, value) {
        this.getCacheStore(strategyName)[key] = value;
    };
    StrategyCacheService.prototype.del = function (strategyName, key) {
        delete this.getCacheStore(strategyName)[key];
    };
    StrategyCacheService.prototype.clear = function (strategyName) {
        this.caches[strategyName] = {};
    };
    StrategyCacheService.prototype.getCacheStore = function (strategyName) {
        if (strategyName in this.caches)
            return this.caches[strategyName];
        return this.caches[strategyName] = {};
    };
    StrategyCacheService.ɵprov = i0.ɵɵdefineInjectable({ factory: function StrategyCacheService_Factory() { return new StrategyCacheService(); }, token: StrategyCacheService, providedIn: "root" });
StrategyCacheService.ɵfac = function StrategyCacheService_Factory(t) { return new (t || StrategyCacheService)(); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(StrategyCacheService, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return []; }, null); })();
    return StrategyCacheService;
}());
export { StrategyCacheService };
