// ngx-webstorage/esm5/lib/services/strategyIndex.js
import { __decorate, __metadata, __param } from "tslib";
import { Subject } from 'rxjs';
import { Inject, Injectable, Optional } from '@angular/core';
import { STORAGE_STRATEGIES } from '../strategies';
import { StorageStrategies } from '../constants/strategy';
import * as i0 from "@angular/core";
import * as i1 from "../strategies/index";
import * as ɵngcc0 from '@angular/core';
export var InvalidStrategyError = 'invalid_strategy';
var StrategyIndex = /** @class */ (function () {
    function StrategyIndex(strategies) {
        this.strategies = strategies;
        this.registration$ = new Subject();
        if (!strategies)
            strategies = [];
        this.strategies = strategies.reverse()
            .map(function (strategy, index, arr) { return strategy.name; })
            .map(function (name, index, arr) { return arr.indexOf(name) === index ? index : null; })
            .filter(function (index) { return index !== null; })
            .map(function (index) { return strategies[index]; });
    }
    StrategyIndex_1 = StrategyIndex;
    StrategyIndex.get = function (name) {
        if (!this.isStrategyRegistered(name))
            throw Error(InvalidStrategyError);
        var strategy = this.index[name];
        if (!strategy.isAvailable) {
            strategy = this.index[StorageStrategies.InMemory];
        }
        return strategy;
    };
    StrategyIndex.set = function (name, strategy) {
        this.index[name] = strategy;
    };
    StrategyIndex.clear = function (name) {
        if (name !== undefined)
            delete this.index[name];
        else
            this.index = {};
    };
    StrategyIndex.isStrategyRegistered = function (name) {
        return name in this.index;
    };
    StrategyIndex.hasRegistredStrategies = function () {
        return Object.keys(this.index).length > 0;
    };
    StrategyIndex.prototype.getStrategy = function (name) {
        return StrategyIndex_1.get(name);
    };
    StrategyIndex.prototype.indexStrategies = function () {
        var _this = this;
        this.strategies.forEach(function (strategy) { return _this.register(strategy.name, strategy); });
    };
    StrategyIndex.prototype.indexStrategy = function (name, overrideIfExists) {
        if (overrideIfExists === void 0) { overrideIfExists = false; }
        if (StrategyIndex_1.isStrategyRegistered(name) && !overrideIfExists)
            return StrategyIndex_1.get(name);
        var strategy = this.strategies.find(function (strategy) { return strategy.name === name; });
        if (!strategy)
            throw new Error(InvalidStrategyError);
        this.register(name, strategy, overrideIfExists);
        return strategy;
    };
    StrategyIndex.prototype.register = function (name, strategy, overrideIfExists) {
        if (overrideIfExists === void 0) { overrideIfExists = false; }
        if (!StrategyIndex_1.isStrategyRegistered(name) || overrideIfExists) {
            StrategyIndex_1.set(name, strategy);
            this.registration$.next(name);
        }
    };
    var StrategyIndex_1;
    StrategyIndex.index = {};
    StrategyIndex.ctorParameters = function () { return [
        { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [STORAGE_STRATEGIES,] }] }
    ]; };
    StrategyIndex.ɵprov = i0.ɵɵdefineInjectable({ factory: function StrategyIndex_Factory() { return new StrategyIndex(i0.ɵɵinject(i1.STORAGE_STRATEGIES, 8)); }, token: StrategyIndex, providedIn: "root" });
    StrategyIndex = StrategyIndex_1 = __decorate([ __param(0, Optional()), __param(0, Inject(STORAGE_STRATEGIES)),
        __metadata("design:paramtypes", [Array])
    ], StrategyIndex);
StrategyIndex.ɵfac = function StrategyIndex_Factory(t) { return new (t || StrategyIndex)(ɵngcc0.ɵɵinject(STORAGE_STRATEGIES, 8)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(StrategyIndex, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return [{ type: Array, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [STORAGE_STRATEGIES]
            }] }]; }, null); })();
    return StrategyIndex;
}());
export { StrategyIndex };
