// ngx-webstorage/esm5/lib/strategies/inMemory.js
import { __decorate, __metadata, __param } from "tslib";
import { of, Subject } from 'rxjs';
import { StrategyCacheService } from '../core/strategyCache';
import { StorageStrategies } from '../constants/strategy';
import { Inject, Injectable } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var InMemoryStorageStrategy = /** @class */ (function () {
    function InMemoryStorageStrategy(cache) {
        this.cache = cache;
        this.keyChanges = new Subject();
        this.isAvailable = true;
        this.name = InMemoryStorageStrategy_1.strategyName;
    }
    InMemoryStorageStrategy_1 = InMemoryStorageStrategy;
    InMemoryStorageStrategy.prototype.get = function (key) {
        return of(this.cache.get(this.name, key));
    };
    InMemoryStorageStrategy.prototype.set = function (key, value) {
        this.cache.set(this.name, key, value);
        this.keyChanges.next(key);
        return of(value);
    };
    InMemoryStorageStrategy.prototype.del = function (key) {
        this.cache.del(this.name, key);
        this.keyChanges.next(key);
        return of(null);
    };
    InMemoryStorageStrategy.prototype.clear = function () {
        this.cache.clear(this.name);
        this.keyChanges.next(null);
        return of(null);
    };
    var InMemoryStorageStrategy_1;
    InMemoryStorageStrategy.strategyName = StorageStrategies.InMemory;
    InMemoryStorageStrategy.ctorParameters = function () { return [
        { type: StrategyCacheService, decorators: [{ type: Inject, args: [StrategyCacheService,] }] }
    ]; };
    InMemoryStorageStrategy = InMemoryStorageStrategy_1 = __decorate([ __param(0, Inject(StrategyCacheService)),
        __metadata("design:paramtypes", [StrategyCacheService])
    ], InMemoryStorageStrategy);
InMemoryStorageStrategy.ɵfac = function InMemoryStorageStrategy_Factory(t) { return new (t || InMemoryStorageStrategy)(ɵngcc0.ɵɵinject(StrategyCacheService)); };
InMemoryStorageStrategy.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: InMemoryStorageStrategy, factory: function (t) { return InMemoryStorageStrategy.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(InMemoryStorageStrategy, [{
        type: Injectable
    }], function () { return [{ type: StrategyCacheService, decorators: [{
                type: Inject,
                args: [StrategyCacheService]
            }] }]; }, null); })();
    return InMemoryStorageStrategy;
}());
export { InMemoryStorageStrategy };
